### PR TITLE
RSDK-2727 - have intel realsense automatically reconnect when device is found again

### PR DESCRIPTION
**Pull request by bhaney**
Previously, unplugging the intel realsense from the raspberry pi would cause the stream to indefinitely freeze on the last seen frame, even if the realsense was plugged back in.

This PR uses the rs2::set_devices_changed_callback function to notice when a Realsense is reconnected to the the raspberry pi, exiting out of the old camera thread, and starting a new camera thread. 

I tested this out on a D435 intel realsense, and the viam-server resumes streaming the cameras after the camera is unplugged and plugged back in.

### CODE REVIEW

File: intel_realsense_grpc.cpp [hunk 1]
```diff
@@ -559,26 +605,70 @@ const PipelineWithProperties startPipeline(const int colorWidth, const int color
...
+    if (!devProps.disableDepth) {
+        auto alignedText = "";
+        if (!devProps.disableColor) {
+            alignedText = " (aligned to color)";
+        }
+        cout << "depth_width:    " << props.depth.width << alignedText << "\n";
+        cout << "depth_height:   " << props.depth.height << alignedText << endl;
+    }
+
+    return make_tuple(pipeline, props);
```
Comment (bhaney) [inline intel_realsense_grpc.cpp:643]: Moved this into the "startPipeline" function so it print again whenever the device is reconnected

File: intel_realsense_grpc.cpp [hunk 2]
```diff
@@ -412,9 +445,23 @@ class RobotServiceImpl final : public RobotService::Service {
 const rs2::align FRAME_ALIGNMENT = RS2_STREAM_COLOR;
 
 void frameLoop(rs2::pipeline pipeline, AtomicFrameSet& frameSet, promise<void>& ready,
-               const bool disableColor, const bool disableDepth, float depthScaleMm) {
+               DeviceProperties& context, float depthScaleMm) {
```
Comment (edaniels) [inline intel_realsense_grpc.cpp:448]:
```suggestion
               DeviceProperties& deviceProps, float depthScaleMm) {
```

File: intel_realsense_grpc.cpp [hunk 3]
```diff
@@ -412,9 +445,23 @@ class RobotServiceImpl final : public RobotService::Service {
...
+        std::lock_guard<std::mutex> lock(context.mutex);
+        context.shouldRun = true;
+        context.isRunning = true;
+    }
     while (true) {
+        {
+            std::lock_guard<std::mutex> lock(context.mutex);
+            if (!context.shouldRun) {
+                pipeline.stop();
+                cout << "[frameLoop] context canceled, exiting thread" << endl;
```
Comment (edaniels) [inline intel_realsense_grpc.cpp:460]:
```suggestion
                cout << "[frameLoop] pipeline stopped exiting thread" << endl;
```

File: intel_realsense_grpc.cpp [hunk 4]
```diff
@@ -559,26 +605,70 @@ const PipelineWithProperties startPipeline(const int colorWidth, const int color
...
+        cout << "depth_width:    " << props.depth.width << alignedText << "\n";
+        cout << "depth_height:   " << props.depth.height << alignedText << endl;
+    }
+
+    return make_tuple(pipeline, props);
+};
+
+void on_device_reconnect(rs2::event_information& info, DeviceProperties& context,
+                         rs2::pipeline pipeline) {
+    if (info.was_added(info.get_new_devices().front())) {
```
Comment (bhaney) [inline intel_realsense_grpc.cpp:648]: This function will only trigger when it sees a new device get connected to the raspberry pi. This only allows one RealSense to every be connected to the raspberry pi, as any new connection will exit out of the current thread, and start a new thread with the first realsense device it finds.

File: intel_realsense_grpc.cpp [hunk 5]
```diff
@@ -559,26 +605,70 @@ const PipelineWithProperties startPipeline(const int colorWidth, const int color
...
+        std::cout << "Device was reconnected, restarting pipeline" << std::endl;
+        {
+            std::lock_guard<std::mutex> lock(context.mutex);
+            context.shouldRun = false;
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
```
Comment (bhaney) [inline intel_realsense_grpc.cpp:654]: This is probably not the best way to do this - i just wanted to give some time to make sure the old thread exited before the new thread gets started. There is an "isRunning" variable in the deviceProperties I am trying to use for this, but coordinating between threads is a little tricky.

File: intel_realsense_grpc.cpp (hunk 4 above)
Comment (edaniels) [inline intel_realsense_grpc.cpp:648]: you can early exist with the negation of this case

**Review by edaniels — APPROVED** [review]
LGTM!

File: intel_realsense_grpc.cpp [hunk 6]
```diff
@@ -559,26 +605,70 @@ const PipelineWithProperties startPipeline(const int colorWidth, const int color
...
+            return;
+        }
+        // Start the camera thread
+        promise<void> ready;
+        thread cameraThread(frameLoop, pipeline, ref(latestFrames), ref(ready), ref(context),
+                            props.depthScaleMm);
+        cout << "waiting for camera frame loop thread to be ready..." << flush;
+        ready.get_future().wait();
+        cout << " ready!" << endl;
+        cameraThread.detach();
```
Comment (bhaney) [inline intel_realsense_grpc.cpp:670]: I am worried about a potential memory leak here, if I missed something obvious

**Pull request merged**